### DISCUSSION TITLE
PGX and FieldDescription

### DISCUSSION THREAD

**Discussion by wshekrota** [jackc/pgx#1733] [Q&A] [answered]
I may have this wrong but did something change in this type? Doc indicates there is a DataTypeName string?
Name seems to work fine but other fields not found.

**d.DataTypeName undefined (type pgconn.FieldDescription has no field or method DataTypeName)**

import "github.com/jackc/pgx/v5"
.
.
rows, err := db.Query(context.Background(), sqCmd)
.
.
// rows is the result of having done a select * from table
.
    Fds := rows.FieldDescriptions()
    var tableFieldNames, tableFieldTypes []string
    for _, d := range Fds {
        tableFieldNames = append(tableFieldNames, d.Name)
        tableFieldTypes = append(tableFieldTypes, d.DataTypeName)
    }

I'm trying to assert data types
Thanks for any ideas.
-Walt

**Comment by jackc** [comment]
> I may have this wrong but did something change in this type? Doc indicates there is a DataTypeName string?

I think you are looking at the v3 docs. That doesn't exist in v4 or v5. https://pkg.go.dev/github.com/jackc/pgx/v5@v5.4.3/pgconn#FieldDescription.

The protocol doesn't directly expose the type name. Only the OID. You may be able to get the `pgtype.Map` from your `Conn` and call https://pkg.go.dev/github.com/jackc/pgx/v5@v5.4.3/pgtype#Map.TypeForOID to get the name (assuming pgx/pgtype knows what it is)

**Comment by wshekrota** [comment, marked as answer]
yeah that goes to v3 what I was looking at  .... I see now you do have newer. Google is my friendemy right.
TypeForOid ok I see that will have to see how to get the oid at first glance that looks like a query which is too far to go I think.
Is the OID available in pgx internals?
Thanks

**Comment by wshekrota** [comment]
Going into the future will FieldDescription name continue to exist as it does now? Can I depend on that or will there be an alternative?
Looks like I see TableOID also...
ok I have a direction thanks.

**Comment by wshekrota** [comment]
I have the DataTypeOID fine but ...
Looks like I have to go back to PGX v3 before I get a Conn like this...
type Conn struct {
	Pid           int32             // backend pid
	SecretKey     int32             // key to use to send a cancel query message to the server
	RuntimeParams map[string]string // parameters that have been reported by the server
	PgTypes       map[Oid]PgType    // oids to PgTypes

	TxStatus byte
	// contains filtered or unexported fields
}
Newer ones say unexported fields in Conn.
Is there any other way I can get this typing to work without a hack :) ?
Thanks.

**Comment by jackc** [comment]
Everything is still available, but usually now via method instead of an exposed field.

https://pkg.go.dev/github.com/jackc/pgx/v5/pgconn#PgConn.PID
https://pkg.go.dev/github.com/jackc/pgx/v5/pgconn#PgConn.SecretKey
https://pkg.go.dev/github.com/jackc/pgx/v5/pgconn#PgConn.ParameterStatus

...

https://pkg.go.dev/github.com/jackc/pgx/v5#Conn.TypeMap gets you the pgtype.Map. From there you can call TypeForOID. It is only a map lookup. It is not a database query.

**Comment by wshekrota** [comment]
I got the oid from the field description then cobbled a switch with some known oids then using reflect.TypeOf in default rapidly exposed the small set of asserts. Adding new oids to case sets as I went.Pgtype.numeric Interface{} Are 2 that came up requiring research.I’ll look through the Conn.TypeMap you have suggested to see if it works better.Thanks for your time in this I have a working postgres to document db converter. Well almost.WaltSent from my iPadOn Sep 11, 2023, at 11:00 AM, Jack Christensen ***@***.***> wrote:﻿
Everything is still available, but usually now via method instead of an exposed field.
https://pkg.go.dev/github.com/jackc/pgx/v5/pgconn#PgConn.PID
https://pkg.go.dev/github.com/jackc/pgx/v5/pgconn#PgConn.SecretKey
https://pkg.go.dev/github.com/jackc/pgx/v5/pgconn#PgConn.ParameterStatus
...
https://pkg.go.dev/github.com/jackc/pgx/v5#Conn.TypeMap gets you the pgtype.Map. From there you can call TypeForOID. It is only a map lookup. It is not a database query.

**Comment by wshekrota** [comment]
﻿﻿Looks like the TypeMap codec method PreferredFormat() returns bool not the int16 format code. I get 1 and 0. Is that even the right method?To convert the column values generated by the next to go values I need to…PlanScan()ThenDecodeValue()?What is the format parameter and how do I provide it?ThanksWaltSent from my iPadOn Sep 11, 2023, at 11:00 AM, Jack Christensen ***@***.***> wrote:﻿
Everything is still available, but usually now via method instead of an exposed field.
https://pkg.go.dev/github.com/jackc/pgx/v5/pgconn#PgConn.PID
https://pkg.go.dev/github.com/jackc/pgx/v5/pgconn#PgConn.SecretKey
https://pkg.go.dev/github.com/jackc/pgx/v5/pgconn#PgConn.ParameterStatus
...
https://pkg.go.dev/github.com/jackc/pgx/v5#Conn.TypeMap gets you the pgtype.Map. From there you can call TypeForOID. It is only a map lookup. It is not a database query.

**Comment by jackc** [comment]
> Looks like the TypeMap codec method PreferredFormat() returns bool not the int16 format code.

https://pkg.go.dev/github.com/jackc/pgx/v5@v5.4.3/pgtype#Map.FormatCodeForOID and all the implementers of PreferredFormat return int16 not bool... 

> I get 1 and 0. 

1 and 0 **are** the underlying values for the binary and text format respectively.

---

> To convert the column values generated by the next to go values I need to…PlanScan()ThenDecodeValue()?

Something like that is possible... but unless you are doing something very unusual you shouldn't need to be working at that low level. You could call `Rows.Values()` to have pgx convert and return all values in the row.
